File: Project4/auto_loot.c
/* Language: C
 * Project 4: Auto-loot
 * CS 450 
 * Author: Amethyst Skye */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <time.h>

/* Purpose: structure for holding information associated with loot item
 * int weight: weight of item
 * int value: value of item
 * char *name: name of item */
typedef struct item{
  int weight;
  int value;
  char *name;
}ITEM;

ITEM items[128];

/* Purpose: Simple function that finds the maximum of two values */
int max(int a, int b){
    if (a > b)
        return (a);
    return(b);
}

/* Purpose: Using an iterative approach, we find the best combination of items
 * to keep in players inventory.
 * int capacity: capacity of bag
 * int item_count: number of items to choose from
 * return: Nothing, but prints contents of optimal knapsack, final weight
 * and final value. */
void auto_loot(int capacity, int item_count){
  int knapsack[item_count + 1][capacity + 1];
  int best_item_count = 0;

  /* Build table with potential knapsack contents */
  for (int i = 0; i < item_count + 1; i++){
    for (int j = 0; j < capacity + 1; j++){
      if (i == 0 || j == 0)
        knapsack[i][j] = 0;
      else if (items[i - 1].weight <= j)
        knapsack[i][j] = max(items[i - 1].value + 
          knapsack[i - 1][j - items[i - 1].weight], 
          knapsack[i - 1][j]);
      else{
        knapsack[i][j] = knapsack[i - 1][j];
        best_item_count++;
      }
    }
  }

  int max_value = knapsack[item_count][capacity],
      max_cap = capacity, 
      final_weight = 0, 
      final_value = knapsack[item_count][capacity];
  ITEM final_knapsack_contents[best_item_count];
  int knap_count = 0;
  
  /* Store results in an array */
  for (int i = item_count; i > 0 && max_value > 0; i--){
    if (max_value == knapsack[i - 1][max_cap])
      continue;
    else{
      final_knapsack_contents[knap_count].name = items[i - 1].name;
      final_knapsack_contents[knap_count].weight = items[i - 1].weight;
      final_knapsack_contents[knap_count].value = items[i - 1].value;
      knap_count++;
      
      max_cap -= items[i - 1].weight;
      max_value -= items[i - 1].value;
      final_weight += items[i - 1].weight;
    }
  }
  for (int i = knap_count; i > 0; i--){
    printf("%s, %d, %d\n", 
            final_knapsack_contents[i - 1].name,
            final_knapsack_contents[i - 1].weight,
            final_knapsack_contents[i - 1].value);
  }
  printf("final weight: %d\n", final_weight);
  printf("final value: %d\n", final_value);
  return;
}

/* Purpose: Reads input from stdin, and uses auto_loot function to print
 * contents of the most optimal loot based on the players current capacity.
 * int argc: number of command line arguments provided by user
 * char* argv[]: command line arguments provided
 * return: 0 on success */
int main(int argc, char* argv[]){
    int capacity, weight, value;
    char buffer[200], name[128];
    int itemCount = 0;

    /* Get bag capacity */
    clock_t start = clock();
    fgets(buffer, 200, stdin);
    sscanf(buffer,"%d", &capacity);

    /* Parse input */
    while (fgets(buffer, 200, stdin) != NULL){
    sscanf(buffer, "%[^;];%d;%d", name, &weight, &value);
      items[itemCount].name = malloc(sizeof(char) * strlen(name) + 1);
      strcpy(items[itemCount].name, name);
      items[itemCount].weight = weight;
      items[itemCount].value = value;
      itemCount++;
    }
    auto_loot(capacity, itemCount);
    clock_t end = clock();
    double total_time = (((double) (end - start)) / CLOCKS_PER_SEC) * 1000000;
    printf("time taken in microseconds: %lf\n", total_time);

    for (int i = 0; i < itemCount; i++)
      free(items[i].name);
    return(0);
}

File: Project4/auto_loot.py
# Python
# Project 4: Auto-loot
# CS 450 
# Author: Amethyst Skye

import sys
import time

# Purpose: Reads input from stdin and assigns data to variables
# return: capacity, name_list, weight_list, value_list
def read_input():
    # Arrays for holding item information
    name_list = []
    weight_list = []
    value_list = []
    
    # Read capacity from stdin
    capacity = int(input())

    # Append item information from stdin to appropriate array
    for line in sys.stdin:
        name,weight,value = line.rstrip().split(";")
        name_list.append(name)
        weight_list.append(int(weight))
        value_list.append(int(value))
    return(capacity, name_list, weight_list, value_list)

# Purpose: Looks at all possible loot combinations and decides items
# which are the most valuable given their weight and the total capacity
# of the bag/knapsack.
# capacity: bag/knapsack capacity
# name: list of item names
# weight: list of item weights
# value: list of item values
# return: nothing, but prints out a list of items which are most desirable
# along with the final weight and value of bag
def auto_loot(capacity, name, weight, value):
    n_items = len(value)
    # table of combinations
    sack = [[0 for i in range(capacity + 1)]
            for j in range(n_items + 1)]
             
    # Build sack[][] table
    for i in range(n_items + 1):
        for j in range(capacity + 1):
            if i == 0 or j == 0:
                sack[i][j] = 0
            elif weight[i - 1] <= j:
                sack[i][j] = max(value[i - 1] + sack[i - 1][j - weight[i - 1]], sack[i - 1][j])
            else:
                sack[i][j] = sack[i - 1][j]
 
    items_list = []
    final_value = sack[n_items][capacity]
    final_weight = 0
    j = capacity
    
    for i in range(n_items, 0, -1):
        if final_value <= 0:
            break

        if final_value == sack[i - 1][j]:
            continue
        else:
 
            # This item is included in the bag
            items_list.append(name[i - 1] + ", " + str(weight[i - 1]) + ", " + str(value[i - 1]))
            final_weight += weight[i - 1]
            final_value -= value[i - 1]
            j -= weight[i - 1]
    
    items_list.reverse()    
    for i in items_list:
        print(i)   
    print("final weight:", final_weight)
    final_value = sack[n_items][capacity]
    print("final value:",final_value)
   
begin = time.time() 
capacity, name_list, weight_list, value_list = read_input()
auto_loot(capacity, name_list, weight_list, value_list)
end = time.time() 
print("time taken in microseconds:", (end - begin) * 1000000)